File: packages/adapters/src/activity.ts
import { z } from "zod";
import {
  API_ACTIVITY_LIMITS,
  attemptEvaluation,
  isApiActivityRisk,
} from "@janitor/core";
import type {
  ApiActivityAssessment,
  ApiActivityContext,
  ApiActivityKey,
  ApiActivityResult,
  ApiActivityStorage,
  VisitorEvaluator,
} from "@janitor/core";
import { createSubjectLinker } from "./subject.js";
import type { SubjectLinkingOptions } from "./subject.js";

const route = z
  .string()
  .max(160)
  .regex(/^(GET|POST|PUT|PATCH|DELETE|HEAD|OPTIONS) \/[A-Za-z0-9_/:.*{}-]*$/);
const keySchema = z
  .object({
    kind: z.enum(["session", "actor"]),
    id: z
      .string()
      .min(1)
      .max(512)
      .refine((v) => !!v.trim()),
  })
  .strict();
const contextSchema = z
  .object({
    key: keySchema,
    route,
    actor: z
      .object({ kind: z.enum(["person", "agent"]), delegated: z.boolean() })
      .strict()
      .optional(),
  })
  .strict()
  .refine((v) => !v.actor || v.key.kind === "actor");
const optionsSchema = z
  .object({
    routes: z
      .array(
        z.object({ route, sensitive: z.boolean().default(false) }).strict(),
      )
      .min(1)
      .max(API_ACTIVITY_LIMITS.routes)
      .refine((v) => new Set(v.map((r) => r.route)).size === v.length),
    windowMs: z.number().int().min(10_000).max(900_000).default(60_000),
    retentionDays: z.number().int().min(1).max(30).default(1),
    minRequests: z.number().int().min(1).max(10_000).default(20),
    evaluationIntervalMs: z
      .number()
      .int()
      .min(5000)
      .max(900_000)
      .default(60_000),
    timeoutMs: z.number().int().min(50).max(5000).default(1500),
    maxInFlight: z.number().int().min(1).max(1024).default(64),
  })
  .strict();
export type ApiActivityOptions = z.input<typeof optionsSchema>;
export type {
  ApiActivityContext,
  ApiActivityKey,
  ApiActivityResult,
  ApiActivityAssessment,
} from "@janitor/core";

/** Optional server telemetry. It never reads a body, URL, header or browser signal. */
export function createApiActivity(
  storage: ApiActivityStorage,
  identity: SubjectLinkingOptions,
  options: ApiActivityOptions,
  evaluator?: VisitorEvaluator,
  evaluatorTimeoutMs = 1200,
) {
  const config = optionsSchema.parse(options);
  const routes = new Map(config.routes.map((r) => [r.route, r]));
  const label = createSubjectLinker(identity);
  const owner = (key: ApiActivityKey) =>
    label(JSON.stringify(["api-activity-v1", key.kind, key.id]));
  let inFlight = 0;
  // A stalled driver continues occupying its slot until it settles. A timeout
  // prevents later evaluation; it must not make room for unbounded stuck work.
  async function limited(
    run: (check: () => void) => Promise<ApiActivityResult>,
  ): Promise<ApiActivityResult> {
    if (inFlight >= config.maxInFlight) return { status: "unavailable" };
    inFlight++;
    let active = true;
    let timer: ReturnType<typeof setTimeout> | undefined;
    const work = Promise.resolve()
      .then(() =>
        run(() => {
          if (!active) throw new Error("Activity deadline");
        }),
      )
      .catch((): ApiActivityResult => ({ status: "unavailable" }))
      .finally(() => {
        inFlight--;
      });
    try {
      return await Promise.race([
        work,
        new Promise<ApiActivityResult>((resolve) => {
          timer = setTimeout(() => {
            active = false;
            resolve({ status: "unavailable" });
          }, config.timeoutMs);
        }),
      ]);
    } finally {
      clearTimeout(timer);
    }
  }
  async function assess(
    context: ApiActivityContext,
    key: string,
    check: () => void,
  ): Promise<ApiActivityResult> {
    const now = Date.now();
    const currentWindow = Math.floor(now / config.windowMs) * config.windowMs;
    // The requested route and verified actor context are part of the cache key.
    // An assessment for one permission context is never reused for another.
    const cacheKey = await label(
      JSON.stringify([
        "api-assessment-v1",
        key,
        context.route,
        context.actor?.kind ?? null,
        context.actor?.delegated ?? null,
        config.windowMs,
        routes.get(context.route)!.sensitive,
      ]),
    );
    check();
    const lease = crypto.randomUUID();
    const expiresAt = now + config.evaluationIntervalMs;
    const claimed = await storage.claim(cacheKey, key, lease, now, expiresAt);
    check();
    if (!claimed) {
      const cached = await storage.cached(cacheKey, now);
      check();
      return cached
        ? { status: "recorded", assessment: { ...cached, cached: true } }
        : { status: "recorded" };
    }
    const rows = await storage.recent(
      key,
      currentWindow - (API_ACTIVITY_LIMITS.windows - 1) * config.windowMs,
      currentWindow,
    );
    check();
    const summary = {
      source: "application-api" as const,
      observedAt: now,
      windowMs: config.windowMs,
      truncated: rows.length > API_ACTIVITY_LIMITS.rows,
      buckets: rows
        .slice(0, API_ACTIVITY_LIMITS.rows)
        .filter((row) => routes.has(row.route)),
    };
    const risk =
      evaluator?.evaluateActivity && summary.buckets.length
        ? await attemptEvaluation(
            () =>
              evaluator.evaluateActivity!({
                activity: summary,
                route: context.route,
                sensitive: routes.get(context.route)!.sensitive,
                ...(context.actor ? { actor: context.actor } : {}),
              }),
            evaluatorTimeoutMs,
            isApiActivityRisk,
          )
        : undefined;
    check();
    const assessment: ApiActivityAssessment = {
      source: "application-api",
      evaluatedAt: now,
      expiresAt,
      summary,
      cached: false,
      risk: risk
        ? { automation: risk.automation, suspicious: risk.suspicious }
        : { automation: 0, suspicious: 0 },
      riskStatus: risk
        ? "evaluated"
        : evaluator?.evaluateActivity
          ? "unavailable"
          : "disabled",
    };
    await storage.save(cacheKey, lease, assessment);
    check();
    return { status: "recorded", assessment };
  }
  const service = {
    /** Record one completed request. Repeated calls count as repeated observations. */
    observe(
      context: ApiActivityContext | undefined,
      outcome: { status: number; durationMs: number },
    ): Promise<ApiActivityResult> {
      if (!context) return Promise.resolve({ status: "skipped" });
      return limited(async (check) => {
        const parsed = contextSchema.parse(context);
        const configured = routes.get(parsed.route);
        if (!configured) return { status: "skipped" };
        const result = z
          .object({
            status: z.number().int().min(100).max(599),
            durationMs: z.number().finite().nonnegative(),
          })
          .strict()
          .parse(outcome);
        const key = await owner(parsed.key);
        check();
        const now = Date.now();
        const row = await storage.increment({
          key,
          route: parsed.route,
          windowStart: Math.floor(now / config.windowMs) * config.windowMs,
          now,
          expiresAt: now + config.retentionDays * 86_400_000,
          status: result.status,
          durationMs: Math.min(
            API_ACTIVITY_LIMITS.maxDurationMs,
            Math.round(result.durationMs),
          ),
        });
        check();
        // Count doubling, first denial and configured sensitive routes trigger
        // assessment attempts. A shared lease/cache and global budget bound AI.
        const ratio = row.requests / config.minRequests;
        const milestone = ratio >= 1 && Number.isInteger(Math.log2(ratio));
        if (
          configured.sensitive ||
          milestone ||
          (row.denied === 1 && (result.status === 401 || result.status === 403))
        )
          return assess(parsed, key, check);
        return { status: "recorded" };
      });
    },
    /** Read/evaluate recent activity explicitly before a sensitive action. */
    assess(context: ApiActivityContext): Promise<ApiActivityResult> {
      return limited(async (check) => {
        const parsed = contextSchema.parse(context);
        if (!routes.has(parsed.route)) return { status: "skipped" };
        const key = await owner(parsed.key);
        check();
        return assess(parsed, key, check);
      });
    },
    /** Web middleware: the caller returns only response; activity stays private. */
    async handle(
      request: Request,
      context: ApiActivityContext | undefined,
      next: (request: Request) => Promise<Response> | Response,
    ): Promise<{ response: Response; activity: ApiActivityResult }> {
      const start = performance.now();
      let response: Response;
      try {
        response = await next(request);
      } catch (error) {
        await service.observe(context, {
          status: 500,
          durationMs: performance.now() - start,
        });
        throw error;
      }
      const activity = await service.observe(context, {
        status: response.status,
        durationMs: performance.now() - start,
      });
      return { response, activity };
    },
    /** Authorize erasure in the application and stop collection before calling. */
    deleteKey: async (key: ApiActivityKey) =>
      storage.deleteKey(await owner(keySchema.parse(key))),
    cleanup: () => storage.cleanup(Date.now(), 100),
  };
  return service;
}

File: packages/elixir/lib/doorman/activity.ex
defmodule Doorman.Activity do
  @moduledoc "Opt-in, bounded API activity on the implementer's server. Never an authorization or identity decision."
  alias Doorman.ActivityStore, as: Store
  @route ~r/^(GET|POST|PUT|PATCH|DELETE|HEAD|OPTIONS) \/[A-Za-z0-9_\/:.*{}-]*$/
  @defaults [
    window_ms: 60_000,
    retention_days: 1,
    min_requests: 20,
    evaluation_interval_ms: 60_000,
    timeout_ms: 1500,
    evaluator: nil
  ]

  def configure(nil, _identity), do: nil

  def configure(opts, identity) do
    if is_nil(identity), do: raise(ArgumentError, "API activity requires identity configuration")

    unless Keyword.keyword?(opts) and
             Enum.all?(Keyword.keys(opts), &(&1 in [:routes | Keyword.keys(@defaults)])),
           do: raise(ArgumentError, "invalid API activity options")

    config = Map.new(Keyword.merge(@defaults, opts))

    unless is_list(config[:routes]) and length(config.routes) in 1..32,
      do: raise(ArgumentError, "API activity requires 1-32 routes")

    routes =
      Map.new(config.routes, fn r ->
        unless is_map(r) and Map.keys(r) -- [:route, :sensitive] == [] and valid_route?(r[:route]) and
                 is_boolean(Map.get(r, :sensitive, false)),
               do: raise(ArgumentError, "invalid API route template")

        {r.route, Map.get(r, :sensitive, false)}
      end)

    if map_size(routes) != length(config.routes), do: raise(ArgumentError, "duplicate API routes")

    for {key, min, max} <- [
          {:window_ms, 10_000, 900_000},
          {:retention_days, 1, 30},
          {:min_requests, 1, 10_000},
          {:evaluation_interval_ms, 5000, 900_000},
          {:timeout_ms, 50, 5000}
        ] do
      value = config[key]

      unless is_integer(value) and value >= min and value <= max,
        do: raise(ArgumentError, "invalid API activity limit")
    end

    unless is_nil(config.evaluator) or is_function(config.evaluator, 1),
      do: raise(ArgumentError, "invalid activity evaluator")

    %{config | routes: routes}
  end

  def observe(c, context, outcome) do
    if is_nil(c.activity) or is_nil(context),
      do: %{"status" => "skipped"},
      else:
        bounded(c, fn ->
          context!(context)

          if Map.has_key?(c.activity.routes, context.route) do
            unless is_map(outcome) and Map.keys(outcome) -- [:status, :duration_ms] == [] and
                     is_integer(outcome[:status]) and outcome.status in 100..599 and
                     is_number(outcome[:duration_ms]) and outcome.duration_ms >= 0,
                   do: raise(ArgumentError, "invalid API outcome")

            owner = owner(c, context.key)

            row =
              Store.increment(
                c,
                owner,
                context.route,
                outcome.status,
                outcome.duration_ms,
                Doorman.now()
              )

            count = row["requests"]

            milestone =
              rem(count, c.activity.min_requests) == 0 and
                power_of_two?(div(count, c.activity.min_requests))

            if c.activity.routes[context.route] or milestone or
                 (row["denied"] == 1 and outcome.status in [401, 403]),
               do: evaluate(c, context, owner),
               else: %{"status" => "recorded"}
          else
            %{"status" => "skipped"}
          end
        end)
  end

  @doc "Assess recent completed activity before a sensitive action; returns private advisory context."
  def assess(c, context) do
    if is_nil(c.activity),
      do: %{"status" => "skipped"},
      else:
        bounded(c, fn ->
          context!(context)

          if Map.has_key?(c.activity.routes, context.route),
            do: evaluate(c, context, owner(c, context.key)),
            else: %{"status" => "skipped"}
        end)
  end

  defp evaluate(c, context, owner) do
    now = Doorman.now()
    actor = context[:actor]

    id =
      Doorman.Identity.label(
        c.identity,
        Jason.encode!([
          "api-assessment-v1",
          owner,
          context.route,
          if(actor, do: actor.kind),
          if(actor, do: actor.delegated),
          c.activity.window_ms,
          c.activity.routes[context.route]
        ])
      )

    lease = Doorman.random_id("activity_")
    expiry = now + c.activity.evaluation_interval_ms

    if Store.claim(c, id, owner, lease, now, expiry) do
      rows = Store.recent(c, owner, now)

      summary = %{
        "source" => "application-api",
        "observedAt" => now,
        "windowMs" => c.activity.window_ms,
        "truncated" => length(rows) > 128,
        "buckets" =>
          rows |> Enum.take(128) |> Enum.filter(&Map.has_key?(c.activity.routes, &1["route"]))
      }

      input = %{
        "activity" => summary,
        "route" => context.route,
        "sensitive" => c.activity.routes[context.route]
      }

      input =
        if actor,
          do: Map.put(input, "actor", %{"kind" => actor.kind, "delegated" => actor.delegated}),
          else: input

      fun =
        cond do
          is_function(c.activity.evaluator, 1) -> fn -> c.activity.evaluator.(input) end
          is_list(c.evaluator) -> fn -> Doorman.Jev.evaluate_activity(input, c.evaluator) end
          true -> nil
        end

      protection =
        c.protection || Doorman.Protection.configure(c.identity ++ [evaluator: [max_calls: 60]])

      result =
        if fun && summary["buckets"] != [],
          do: Doorman.Protection.evaluate(%{c | protection: protection}, fun, &valid_risk?/1),
          else: nil

      risk =
        case result do
          {:ok, r} -> if valid_risk?(r), do: Map.take(r, ~w(automation suspicious))
          _ -> nil
        end

      assessment = %{
        "source" => "application-api",
        "evaluatedAt" => now,
        "expiresAt" => expiry,
        "summary" => summary,
        "cached" => false,
        "risk" => risk || %{"automation" => 0, "suspicious" => 0},
        "riskStatus" =>
          cond do
            risk -> "evaluated"
            fun -> "unavailable"
            true -> "disabled"
          end
      }

      Store.save(c, id, lease, assessment)
      %{"status" => "recorded", "assessment" => assessment}
    else
      case Store.cached(c, id, now) do
        nil -> %{"status" => "recorded"}
        cached -> %{"status" => "recorded", "assessment" => Map.put(cached, "cached", true)}
      end
    end
  end

  def valid_risk?(value) when is_map(value),
    do:
      Enum.all?(~w(automation suspicious), fn k ->
        is_number(value[k]) and value[k] >= 0 and value[k] <= 1
      end)

  def valid_risk?(_), do: false

  def delete_key(c, key) do
    key!(key)
    Store.delete(c, owner(c, key))
  end

  def cleanup(c), do: Store.cleanup(c)

  defp bounded(c, fun) do
    case Doorman.Bounded.run(fun, c.activity.timeout_ms) do
      {:ok, result} -> result
      _ -> %{"status" => "unavailable"}
    end
  end

  defp owner(c, key),
    do: Doorman.Identity.label(c.identity, Jason.encode!(["api-activity-v1", key.kind, key.id]))

  defp valid_route?(value),
    do: is_binary(value) and byte_size(value) <= 160 and Regex.match?(@route, value)

  defp key!(key) do
    unless is_map(key) and Map.keys(key) -- [:kind, :id] == [] and
             key[:kind] in ["session", "actor"] and is_binary(key[:id]) and
             byte_size(key.id) in 1..512 and String.trim(key.id) != "",
           do: raise(ArgumentError, "invalid API activity key")
  end

  defp context!(context) do
    unless is_map(context) and Map.keys(context) -- [:key, :route, :actor] == [] and
             valid_route?(context[:route]),
           do: raise(ArgumentError, "invalid API activity context")

    key!(context[:key])

    if actor = context[:actor] do
      unless context.key.kind == "actor" and is_map(actor) and
               Map.keys(actor) -- [:kind, :delegated] == [] and
               actor[:kind] in ["person", "agent"] and is_boolean(actor[:delegated]),
             do: raise(ArgumentError, "invalid verified API actor")
    end
  end

  defp power_of_two?(1), do: true
  defp power_of_two?(n) when n > 1 and rem(n, 2) == 0, do: power_of_two?(div(n, 2))
  defp power_of_two?(_), do: false
end
